packages/typescript-estree/tests/lib/semanticInfo.test.ts

```diff
@@ -172,6 +172,7 @@ describe('semanticInfo', () => {
     ).declarations[0].init!;
     const tsBinaryExpression =
       parseResult.services.esTreeNodeToTSNodeMap.get(binaryExpression);
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- intentionally asserting that the runtime value matches the types.
```

Comment:
🤔 I feel like we could just use an `as` for this file's asserts... they're roughly doing the same thing.

Reply:
well... no. These don't change the types, _that's why they're flagged_ 😀 

And this is exactly one of the cases where asserting that the runtime values match the types is warranted; the `esTreeNodeToTSNodeMap` and its inverse can well have unmapped values, see also https://github.com/typescript-eslint/typescript-eslint/issues/9268#issuecomment-2148437137 for a real bug caused by this.

However, I did notice that some of the assertions were actually redundant, and others could be replaced with an ordinary `expect(x).toBeDefined()`, which really just hides the unnecessariness from the rule. 

Made some changes 👍